### DISCUSSION TITLE
v13-beta server components not working as I expected?

### DISCUSSION THREAD

**Discussion by NashJames** [vercel/next.js#43788] [Help] [answered]
Wasn't certain whether this should be asked here or in React, but as there's a bunch of additional stuff going on under the hood I figured here might be better.

I've got two very simple files in the app directory. The first is a very basic fetch in a Server Component:
```tsx
export default async function InitialFetch() {
  const res = await fetch('https://swapi.dev/api/people/1/', {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
    next: { revalidate: 60 * 60 * 24 },
  })

  if (!res.ok) return res.status.toString()

  return JSON.stringify(await res.json(), null, 2)
}
```

The other a relatively empty XML file:
```tsx
'use client'
import { useEffect, useState } from 'react'
import InitialFetch from './InitialFetch'

export default function FetchContainer() {
  const [json, setJson] = useState<string>('')

  // Runs on page load
  useEffect(() => {
    const fetchInitialState = async () => setJson(await InitialFetch())
    fetchInitialState()
  }, [])

  return <p>{json}</p>
}
```

The purpose of the useEffect is because the XML function cannot be async and it breaks the rules of hooks to include the `await IntialFetch()` inside a useState default, but I do not believe it should have any delay on page load?

There's a good couple seconds before the JSON renders.

As I understood the concept, the server would render `InitialFetch()` and have it sitting ready to be sent to the client as just a dump of json without the need to execute the fetch on page load? But the delay leads me to think otherwise.

I'm assuming I have misunderstood something? Thanks for any help

**Comment by icyJoseph** [comment, marked as answer]
I think you've got a few concepts mixed up.

With Server Components you can await the data on the server. The idea is to **fetch and then render**, rather than **render and then fetch.**

```jsx
import InitialFetch from './InitialFetch'

// this could just be the `page.tsx` component
// or be directly used by one
export default async function FetchContainer() {
  const data = await InitialFetch();
  return <p>{data}</p>
}
```

So everything can just happen server side. If you need interactivity then you can do that on a child of `FetchContainer`, which should use `use client` at the top of the file where it is defined.

**Reply by NashJames** [reply]
I appreciate the reply. This has helped me make more sense of the other discussion several days ago on the same topic, which you also helpfully answered. Thanks for that.

Looks like there's a bit more work to be done in React 18 before the concept becomes properly useful, even on small projects. Thanks for the understanding. Will be keeping an eye on it.
